src/util/area_store.h

```diff
@@ -36,141 +36,152 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
 class AreaStore {
-protected:
-	// TODO change to unordered_map when we can
-	std::map<u32, Area> areas_map;
-	void invalidateCache();
-	virtual void getAreasForPosImpl(std::vector<Area *> *result, v3s16 pos) = 0;
-	bool cache_enabled; // don't write to this from subclasses, only read.
 public:
-	virtual void insertArea(const Area &a) = 0;
+	AreaStore() :
+		m_cache_enabled(true),
+		m_cacheblock_radius(64),
+		m_res_cache(1000, &cacheMiss, this),
+		m_next_id(0)
+	{}
+
+	virtual ~AreaStore() {}
+
+	static AreaStore *getOptimalImplementation();
+
 	virtual void reserve(size_t count) {};
+	size_t size() const { return areas_map.size(); }
+
+	/// Add an area to the store.
+	/// Updates the area's ID.
+	virtual void insertArea(Area &a) = 0;
+
+	/// Removes an area from the store by ID.
+	/// @return Whether the area was in the store and removed.
 	virtual bool removeArea(u32 id) = 0;
+
+	/// Finds areas that the passed position is contained in.
+	/// Stores output in passed vector.
 	void getAreasForPos(std::vector<Area *> *result, v3s16 pos);
+
+	/// Finds areas that are completely contained inside the area defined
+	/// by the passed edges.  If @p accept_overlap is true this finds any
+	/// areas that intersect with the passed area at any point.
 	virtual void getAreasInArea(std::vector<Area *> *result,
 		v3s16 minedge, v3s16 maxedge, bool accept_overlap) = 0;
 
-#if 0
-	// calls a passed function for every stored area, until the
-	// callback returns true. If that happens, it returns true,
-	// if the search is exhausted, it returns false
-	virtual bool forEach(bool (*callback)(void *args, Area *a), void *args) const = 0;
-#endif
-
-	virtual ~AreaStore()
-	{}
-
-	AreaStore() :
-		cache_enabled(true),
-		m_cacheblock_radius(64),
-		m_res_cache(1000, &cacheMiss, this),
-		m_highest_id(0)
-	{
-	}
-
+	/// Sets cache parameters.
 	void setCacheParams(bool enabled, u8 block_radius, size_t limit);
 
-	u32 getFreeId(v3s16 minedge, v3s16 maxedge);
+	/// Returns a pointer to the area coresponding to the passed ID,
+	/// or NULL if it doesn't exist.
 	const Area *getArea(u32 id) const;
-	u16 size() const;
+
 #if 0
-	bool deserialize(std::istream &is);
+	typedef bool (*ForEachCallback)(const Area *a, void *arg);
+	/// Calls a passed function for every stored area, until the
+	/// callback returns true.  If that happens, it returns true,
+	/// if the search is exhausted, it returns false.
+	virtual bool forEach(ForEachCallback, void *arg=NULL) const = 0;
+
 	void serialize(std::ostream &is) const;
+	bool deserialize(std::istream &is);
 #endif
+
+protected:
+	/// Invalidates the getAreasForPos cache.
+	/// Call after adding or removing an area.
+	void invalidateCache();
+
+	/// Implementation of getAreasForPos.
+	/// getAreasForPos calls this if the cache is disabled.
+	virtual void getAreasForPosImpl(std::vector<Area *> *result, v3s16 pos) = 0;
+
+	/// Returns the next area ID and increments it.
+	u32 getNextId() { return m_next_id++; }
```

Comment:
It will only overflow if you insert more than four billion areas into an `AreaStore`, which is FAR more than it's designed to hold.  I could add an assertion if you really think that's something that needs to be checked though.